### PR TITLE
Release/4.2

### PR DESCRIPTION
**Pull request by kyranjamie**
> [Download the latest build](https://github.com/blockstack/stacks-wallet/actions/runs/809638400)

**Updates to Electron and all packages**
I'll rebase these commits before merging.

I've been unable to upgrade `electron-rebuild` to the latest version. As of [`2.0.3` a new error ](https://github.com/electron/electron-rebuild/commit/89c013bed803f45419847a6b8f98f4722e6607f2#diff-7253ac06d34f5024e26d50ece97b7ad6fb0feda850d39231f51d5a53a7c1bd1cR185) gets thrown breaking the build.

**QA table**

|                          	| Windows 10 	| MacOS 	| Linux (Fedora)	|
|--------------------------	|---------	|-------	|-------	|
| Creating wallet Software 	|     ✅     	| ✅     	|       ✅	|
| Creating wallet Ledger   	|        ✅     	|  ✅     	| ✅     	|
| Restore wallet           	|       ✅   	| ✅     	| ✅     	|
| Sending tx software      	|      ✅    	| ✅     	|       ✅	|
| Sending tx ledger        	|         ✅    	|     ✅  	| ✅     	|
| Delegate Software        	|       ✅   	|       ✅	|    ✅   	|
| Delegate Ledger          	|       ✅      	|   ✅     	| ✅     	|
| Revoke Software          	|        ✅  	|    ✅   	|    ✅   	|
| Revoke Ledger            	|        ✅     	|    ✅  	| ✅     	|

### CODE REVIEW

File: app/main.dev.ts [hunk 1]
```diff
@@ -16,16 +16,23 @@
...
 import path from 'path';
-import { app, BrowserWindow, clipboard, ipcMain, Menu, session } from 'electron';
-import Store from 'electron-store';
+import { app, BrowserWindow, ipcMain, session } from 'electron';
 import windowState from 'electron-window-state';
 import contextMenu from 'electron-context-menu';
+// import installExtension, {
+//   REACT_DEVELOPER_TOOLS,
+//   REDUX_DEVTOOLS,
+// } from 'electron-devtools-installer';
```
Comment (aulneau) [inline app/main.dev.ts:25]: remove

File: app/main.dev.ts [hunk 2]
```diff
@@ -40,31 +47,23 @@ if (process.env.NODE_ENV === 'production') {
...
   if (process.env.NODE_ENV === 'development' || process.env.DEBUG_PROD === 'true') {
-    await installExtensions();
+    await app.whenReady();
+    // await installExtension([REDUX_DEVTOOLS, REACT_DEVELOPER_TOOLS], {
+    //   loadExtensionOptions: { allowFileAccess: true },
+    //   forceDownload: true,
+    // } as any)
+    //   .then(name => console.log(`Added Extension:  ${name}`))
+    //   // .catch(err => console.log('An error occurred: ', err))
+    //   .finally(() => require('electron-debug')());
```
Comment (aulneau) [inline app/main.dev.ts:62]: remove

File: configs/webpack.config.base.js [hunk 3]
```diff
@@ -41,18 +53,18 @@ export default {
...
-    process: false,
-    child_process: 'empty',
-  },
+  // node: {
+  //   __dirname: false,
+  //   // fs: false,
+  //   Buffer: false,
+  //   process: false,
+  //   // child_process: false,
+  // },
```
Comment (aulneau) [inline configs/webpack.config.base.js:62]: remove

File: configs/webpack.config.renderer.dev.babel.js [hunk 4]
```diff
@@ -47,12 +50,19 @@ export default merge(baseConfig, {
...
+  // entry: [
+  //   'core-js',
+  //   'regenerator-runtime/runtime',
+  //   ...(process.env.PLAIN_HMR ? [] : ['react-hot-loader/patch']),
+  //   `webpack-dev-server/client?http://localhost:${port}/`,
+  //   'webpack/hot/only-dev-server',
+  //   require.resolve('../app/polyfill.ts'),
+  //   require.resolve('../app/index.tsx'),
+  // ],
+
```
Comment (aulneau) [inline configs/webpack.config.renderer.dev.babel.js:62]: remove

File: configs/webpack.config.renderer.dev.babel.js [hunk 5]
```diff
@@ -64,6 +74,18 @@ export default merge(baseConfig, {
...
   module: {
     rules: [
+      {
+        test: /\.[jt]sx?$/,
+        exclude: /node_modules/,
+        use: [
+          {
+            loader: require.resolve('babel-loader'),
+            options: {
+              plugins: [require.resolve('react-refresh/babel')].filter(Boolean),
```
Comment (aulneau) [inline configs/webpack.config.renderer.dev.babel.js:84]: don't need `filter(Boolean)`

File: configs/webpack.config.renderer.prod.babel.js [hunk 6]
```diff
@@ -8,16 +8,18 @@ import { BundleAnalyzerPlugin } from 'webpack-bundle-analyzer';
...
 
+import baseConfig, { defaultNodePolyfillsForRenderer } from './webpack.config.base';
+
 CheckNodeEnv('production');
 DeleteSourceMaps();
 
 // eslint-disable-next-line import/no-default-export
 export default merge(baseConfig, {
-  devtool: process.env.DEBUG_PROD === 'true' ? 'source-map' : 'none',
+  // devtool: process.env.DEBUG_PROD === 'true' ? 'source-map' : 'none',
```
Comment (aulneau) [inline configs/webpack.config.renderer.prod.babel.js:22]: remove

File: configs/webpack.config.renderer.prod.babel.js [hunk 7]
```diff
@@ -34,13 +36,14 @@ export default merge(baseConfig, {
     path: path.join(__dirname, '..', 'app/dist'),
     publicPath: './dist/',
     filename: 'renderer.prod.js',
-    libraryTarget: 'var',
+    // libraryTarget: 'var',
```
Comment (aulneau) [inline configs/webpack.config.renderer.prod.babel.js:39]: remove

File: app/utils/disk-store.ts [hunk 8]
```diff
@@ -24,27 +24,27 @@ interface LedgerWallet {
 export type DiskStore = LedgerWallet | SoftwareWallet;
 
 export const persistEncryptedMnemonic = (encryptedMnemonic: string) => {
-  void api.store.set(StoreIndex.EncryptedMnemonic, encryptedMnemonic);
+  void main.store.set(StoreIndex.EncryptedMnemonic, encryptedMnemonic);
```
Comment (aulneau) [inline app/utils/disk-store.ts:27]: why are these promises ignored?

File: app/utils/external-links.ts [hunk 9]
```diff
@@ -4,7 +4,7 @@ import { EXPLORER_URL, NETWORK } from '@constants/index';
 
 export async function openExternalLink(url: string) {
   if (!isWebUri(url)) return;
-  return api.openExternalLink(url);
+  return main.openExternalLink(url);
```
Comment (aulneau) [inline app/utils/external-links.ts:7]: simpler?
```ts
return isWebUri(url) && main.openExternalLink(url)
  ```

File: configs/webpack.config.base.js [hunk 10]
```diff
@@ -9,9 +9,21 @@ import ExtendedDefinePlugin from 'extended-define-webpack-plugin';
...
+  zlib: require.resolve('browserify-zlib'),
+  assert: require.resolve('assert'),
+  fs: false,
+  perf_hooks: false,
+};
+
 // eslint-disable-next-line import/no-default-export
 export default {
-  target: 'web',
+  // target: 'web',
```
Comment (aulneau) [inline configs/webpack.config.base.js:26]: remove

**Review by aulneau — COMMENTED** [review]
looks pretty good! nothing too much of note, just some comments left in place that can be removed. Also maybe you're not ready for a review, if so, disregard my comments :)

next step: use esbuild 😎

File: app/utils/disk-store.ts (hunk 8 above)
Comment (hstove) [reply app/utils/disk-store.ts:27]: No need to block UI?

File: app/main.dev.ts (hunk 2 above)
Comment (kyranjamie) [reply app/main.dev.ts:62]: This is a frustrating issue. To load in redux devtools you need to add this, but it throws some runtime errors in dev mode.

Comment (kyranjamie) [reply app/main.dev.ts:62]: Tidied up with comment and issue ref

File: app/utils/disk-store.ts (hunk 8 above)
Comment (kyranjamie) [reply app/utils/disk-store.ts]: Good spot, it was like this pre-using preload.js set up, fixed.

File: app/store/keys/keys.actions.ts [hunk 11]
```diff
@@ -83,9 +83,9 @@ export function setSoftwareWallet({ password, history }: SetSoftwareWallet) {
     const encryptedMnemonic = await encryptMnemonic({ derivedKeyHash, mnemonic });
     const rootNode = await deriveRootKeychainFromMnemonic(mnemonic);
     const { address } = deriveStxAddressKeychain(rootNode);
-    persistStxAddress(address);
-    persistSalt(salt);
-    persistEncryptedMnemonic(encryptedMnemonic);
+    await persistStxAddress(address);
+    await persistSalt(salt);
+    await persistEncryptedMnemonic(encryptedMnemonic);
```
Comment (aulneau) [inline app/store/keys/keys.actions.ts:88]:
```suggestion
    await Promise.all([persistStxAddress(address), persistSalt(salt), persistEncryptedMnemonic(encryptedMnemonic)]);
```

File: app/store/keys/keys.actions.ts [hunk 12]
```diff
@@ -40,10 +40,10 @@ interface SetLedgerAddress {
...
 }
 export function setLedgerWallet({ address, publicKey, onSuccess }: SetLedgerAddress) {
-  return (dispatch: Dispatch) => {
-    persistStxAddress(address);
-    persistPublicKey(publicKey);
-    persistWalletType('ledger');
+  return async (dispatch: Dispatch) => {
+    await persistStxAddress(address);
+    await persistPublicKey(publicKey);
+    await persistWalletType('ledger');
```
Comment (aulneau) [inline app/store/keys/keys.actions.ts:46]:
```suggestion
    await Promise.all([persistStxAddress(address), persistPublicKey(publicKey), persistWalletType('ledger')]);
```

File: app/modals/reset-wallet/reset-wallet-modal.tsx [hunk 13]
```diff
@@ -31,7 +31,7 @@ export const ResetWalletModal: FC<ResetWalletModalProps> = ({ isOpen, onClose })
     cancelBtnRef.current?.focus();
     timer.current = window.setTimeout(() => {
       void clearDiskStorage();
-      void api.reloadApp();
+      void main.reloadApp();
```
Comment (aulneau) [inline app/modals/reset-wallet/reset-wallet-modal.tsx:34]: we we want to reload app even if `clearDiskStorage` might still have a pending promise?

File: app/pages/app.tsx [hunk 14]
```diff
@@ -97,25 +97,27 @@ export const App: FC = ({ children }) => {
...
-        ) {
-          await safeAwait(delegationStatus.refetch());
-        }
-        if (newTx.data.tx_status !== 'success') return;
-        dispatch(addNewTransaction(newTx.data));
-        dispatch(pendingTransactionSuccessful(newTx.data));
-      });
+      try {
+        client = await connectWebSocketClient(urljoin(wsUrl.toString(), 'extended', 'v1', 'ws'));
+        if (!address) return;
```
Comment (aulneau) [inline app/pages/app.tsx:102]:
```suggestion
        if (!address) return;
        client = await connectWebSocketClient(urljoin(wsUrl.toString(), 'extended', 'v1', 'ws'));
```

File: app/modals/reset-wallet/reset-wallet-modal.tsx (hunk 13 above)
Comment (kyranjamie) [reply app/modals/reset-wallet/reset-wallet-modal.tsx]: We could mostly get away with this, as it's a separate process handling the delete (of the file on disk). Nonetheless you're right, there could be a race condition here.

**Comment by kyranjamie** [comment]
@aulneau any ideas why the builds failing with esbuild?

**Comment by markmhx** [comment]
I ran into a nonce issue when trying to delegate in software mode:

image
image

**Comment by markmhx** [comment]
It's a small detail, but I believe before it would say "Revoking delegation" on the right while pending and now it remains as "Revoke delegation":

image

**Comment by markmhx** [comment]
The new build opens in dark mode for me by default. @jasperjansz is that intended or do you prefer we open in light mode (my personal preference at least)?

image

**Comment by markmhx** [comment]
The "Reveal address" button here looks thinner than expected given the available room:

image

**Comment by markmhx** [comment]
When I've submitted a transaction for sending then re-open the "Send" dialog while it's pending, I see the pending password form still:

image

**Comment by aulneau** [comment]
> The new build opens in dark mode for me by default. @jasperjansz is that intended or do you prefer we open in light mode (my personal preference at least)?
> 
> image

Is your system in dark mode? My understanding is it defaults to system preference

**Comment by markmhx** [comment]
The icon for "This transaction cannot be reversed" is showing up as more faded than the other two here at the bottom of the Stacking form in dark mode:

image

**Comment by markmhx** [comment]
> Is your system in dark mode? My understanding is it defaults to system preference

Nope, default light mode on MacOS.

**Comment by markmhx** [comment]
This is a nitpick we can ignore, but the circle looks very faint in dark mode. Do we want to make it a bit more noticeable?

image

**Comment by markmhx** [comment]
Dark mode is otherwise looking good! I haven't found any other issues yet 🌚

**Comment by kyranjamie** [comment]
Dodgy revoke button margins 
image

**Comment by kyranjamie** [comment]
Dark mode on Windows looks a bit funny, toolbars etc aren't instructed to change. https://www.electronjs.org/docs/tutorial/dark-mode

**Comment by andresgalante** [comment]
I wouldn't block the release because of dark mode in windows looking odd.

Since this is a short week for @aulneau, I'd say that if this is something that we can fix before the release let's do it, else let's create an issue to track it and do it later. What do you think?

**Comment by kyranjamie** [comment]
@markmhx @hstove @aulneau Rebased commits, we're ready go on this one when approved.

**Review by markmhx — APPROVED** [review]

**Pull request merged**